pkg/applicationserver/io/packages/redis/registry.go

```diff
@@ -0,0 +1,251 @@
+	}
+	return dst, dst.SetFields(src, paths...)
+}
+
+// ApplicationPackagesRegistry is a Redis application packages registry.
+type ApplicationPackagesRegistry struct {
+	Redis *ttnredis.Client
+}
+
+func (r *ApplicationPackagesRegistry) devKey(devUID string) string {
+	return r.Redis.Key("uid", devUID)
+}
+
+func (r *ApplicationPackagesRegistry) fPortStr(fPort uint32) string {
+	return strconv.FormatUint(uint64(fPort), 16)
+}
+
+func (r *ApplicationPackagesRegistry) associationKey(devUID string, fPort string) string {
+	return r.Redis.Key("uid", devUID, fPort)
+}
+
+func (r *ApplicationPackagesRegistry) makeAssociationKeyFunc(devUID string) func(port string) string {
+	return func(port string) string {
+		return r.associationKey(devUID, port)
+	}
+}
+
+// Get implements applicationpackages.AssociationRegistry.
+func (r ApplicationPackagesRegistry) Get(ctx context.Context, ids ttnpb.ApplicationPackageAssociationIdentifiers, paths []string) (*ttnpb.ApplicationPackageAssociation, error) {
+	pb := &ttnpb.ApplicationPackageAssociation{}
+	if err := ttnredis.GetProto(r.Redis, r.associationKey(unique.ID(ctx, ids.EndDeviceIdentifiers), r.fPortStr(ids.FPort))).ScanProto(pb); err != nil {
+		return nil, err
+	}
+	return applyAssociationFieldMask(nil, pb, appendImplicitAssociationGetPaths(paths...)...)
+}
+
+// List implements applicationpackages.AssociationRegistry.
+func (r ApplicationPackagesRegistry) List(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, paths []string) ([]*ttnpb.ApplicationPackageAssociation, error) {
+	var pbs []*ttnpb.ApplicationPackageAssociation
+	devUID := unique.ID(ctx, ids)
+	devKey := r.devKey(devUID)
+	err := r.Redis.Watch(func(tx *redis.Tx) error {
+		tx.Pipelined(func(p redis.Pipeliner) error {
+			var offset, limit int64
+			if limit, offset = limitAndOffsetFromContext(ctx); limit != 0 {
+				if err := countTotal(ctx, devKey, p); err != nil {
+					return err
+				}
+			}
+			return ttnredis.FindProtosPaginated(r.Redis, devKey, r.makeAssociationKeyFunc(devUID), offset, limit).Range(func() (proto.Message, func() (bool, error)) {
+				pb := &ttnpb.ApplicationPackageAssociation{}
+				return pb, func() (bool, error) {
+					pb, err := applyAssociationFieldMask(nil, pb, appendImplicitAssociationGetPaths(paths...)...)
+					if err != nil {
+						return false, err
+					}
+					pbs = append(pbs, pb)
+					return true, nil
+				}
+			})
+		})
+		return nil
+	}, devKey)
+	if err != nil {
+		return nil, err
+	}
+	return pbs, nil
+}
+
+// Set implements applicationpackages.AssociationRegistry.
+func (r ApplicationPackagesRegistry) Set(ctx context.Context, ids ttnpb.ApplicationPackageAssociationIdentifiers, gets []string, f func(*ttnpb.ApplicationPackageAssociation) (*ttnpb.ApplicationPackageAssociation, []string, error)) (*ttnpb.ApplicationPackageAssociation, error) {
+	devUID := unique.ID(ctx, ids.EndDeviceIdentifiers)
+	fPort := r.fPortStr(ids.FPort)
+	ik := r.associationKey(devUID, fPort)
```

Comment:
What does `ik` stand for here?
In e.g. JS registry `ik` stands for "ID Key"; `uk` stands for "UID Key".
Maybe `ak` for "association key"?